### ISSUE TITLE
(package) dependencies not correctly handled with nix

### ISSUE THREAD

**Issue by emillon** [ocaml/dune#6623]
Hi,

The easiest way to see this issue is to run `make test` under `nix develop`.

This produces a bunch of issues like:

```
/nix/store/jzxyplk5jkqdm53l7rk46129x3kwz1wz-mdx-2.1.0/lib/ocaml/4.14.0/site-lib/mdx/../../doc/mdx/README.md
-> required by alias test/blackbox-tests/test-cases/mdx-stanza/runtest
```

This is caused by an expansion of `(deps (package mdx))` which expands to a dependency of each installed file (in each installed section):

https://github.com/ocaml/dune/blob/f244a6e22a79447c8681325eacd87549000b45d9/src/dune_rules/dep_conf_eval.ml#L169-L177

For some reason, the relative paths built here do not correspond to existing files. Possibly this is due to how opam-nix sets up directories?

https://github.com/tweag/opam-nix/blob/429d84f387d8b5def0aeb1e229e9144a2e33baf5/src/builder.nix#L98

**Comment by emillon** [comment]
(switching to `paths_existing` make the errors go away but of course it's probably not the right fix)

**Comment by rgrinberg** [comment]
Yes, it's defo not the right fix. @anmonteiro any idea on how to resolve this? It's a shame that dune-package files don't work correctly in nix.

**Comment by emillon** [comment]
After debugging a bit:
- the only sections for which the path exist are `lib` and `libexec` (which correspond to the same dir)
- `dune-package` files look like:

      (sections (lib .) (libexec .) (bin ../../bin) (doc ../../doc/mdx))

So that might be something with the way we encode `dune-package` files.

**Comment by emillon** [comment]
Whereas opam-nix sets these:

```nix
    {
          prefix = "$out";
          bin = "$out/bin";
          sbin = "$out/bin";
          lib = "$out/lib/ocaml/\${opam__ocaml__version}/site-lib";
          man = "$out/share/man";
          doc = "$out/share/doc";
          share = "$out/share";
          etc = "$out/etc";
    }
```

So it does not seem compatible.

**Comment by emillon** [comment]
I think I might have an idea about what's going on... dune creates `dune-package` in `_build` depending on the roots it knows about as the package is built. But the actual paths are only known at install time. So the relative paths created by `Path.reach` are not correct anymore when the `dune-package` file is installed.

The problem is that I don't know how to test that in nix. I'd like to build the equivalent of `nix develop` but where `dune_3` (the dune used to build all other packages) is sourced from another place. How can I do that?

**Comment by emillon** [comment]
This is interesting, on that flake the paths in dune-package are absolute (I think that's what I would have expected), but still do not correspond to the installed locations:

```
(sections
 (lib
  /nix/store/gl5w7b5r8w9819qlv289fd1yhzq2z424-ocaml4.14.0-integers-0.5.1/lib/ocaml/4.14.0/site-lib/integers)
 (libexec
  /nix/store/gl5w7b5r8w9819qlv289fd1yhzq2z424-ocaml4.14.0-integers-0.5.1/lib/ocaml/4.14.0/site-lib/integers)
 (doc
  /nix/store/gl5w7b5r8w9819qlv289fd1yhzq2z424-ocaml4.14.0-integers-0.5.1/doc/integers)
 (stublibs
  /nix/store/gl5w7b5r8w9819qlv289fd1yhzq2z424-ocaml4.14.0-integers-0.5.1/lib/ocaml/4.14.0/site-lib/stublibs))
```

but doc is installed under `/nix/store/gl5w7b5r8w9819qlv289fd1yhzq2z424-ocaml4.14.0-integers-0.5.1/share/doc`.

**Comment by emillon** [comment]
@Alizter I'm making changes to dune, but when I'm re-running `nix develop .#slim` it's entering the shell immediately without rebuilding the flake. Is there a way to force it to use the new dune?

**Comment by emillon** [comment]
`dune install` has a hook that rewrites paths to absolute ones but I don't think that `opam install` does anything like that. So I'm wondering how this all works and in particular why the slim shell creates absolute paths.

**Comment by emillon** [comment]
ah I saw this in `nix-log`:

    moving /nix/store/a1wm2zyvc8nr0cvnrgpywggghazx6khl-ocaml4.14.0-cinaps-0.15.1/doc to /nix/store/a1wm2zyvc8nr0cvnrgpywggghazx6khl-ocaml4.14.0-cinaps-0.15.1/share/doc

I'll check where this comes from

**Comment by emillon** [comment]
Looks like it could be https://github.com/NixOS/nixpkgs/blob/nixpkgs-unstable/pkgs/build-support/setup-hooks/move-docs.sh - how is it imported?

**Comment by emillon** [comment]
ok, my strategy is to add a preFixupPhase that will substitute in place doc/pkg by share/doc/pkg in `dune-package`.

**Comment by anmonteiro** [comment]
> preFixupPhase

I'd expect this to be a `postPatch` or `prePatch`

**Comment by emillon** [comment]
> note Nix will use dune install

that's for a dune build though? I was under the impression that opam-nix uses `buildOpamProject'`, but I'm not sure what's going on with all these nix files. 

> I'd expect this to be a postPatch or prePatch

`patch` operates before building, but dune-package files are created during the build.

I managed to get this to work using the following piece of code in `builder.nix` in `opam-nix`:

```nix
{       # new in preFixupPhase
        patchDunePackagePhase = ''
          dune_package="$out"/lib/ocaml/''${opam__ocaml__version}/site-lib/''${pname}/dune-package
          if [[ -f "$dune_package" ]] ; then
            substituteInPlace "$dune_package" --replace ../.. ../../../../..
          fi
        '';

        forceShare = [ "info" ];
}
```

This changes the relative path to the right level, and removes moving from `doc` to `share/doc` (I guess it would be possible to remove the `forceShare` part and patch the `dune-package` file more precisely by adding `share` at the right place.

This fixes all the issues with "file not available", but still I think this shows an issue with `dune-package` files in general: we're computing relative directories that are correct under `_build` but can become invalid after installation. So I'm not sure what is the right fix.

**Comment by anmonteiro** [comment]
> that's for a dune build though? 

My bad. I don't think I can help with `buildOpamProject'`

**Comment by anmonteiro** [comment]
A few months later, there's a solution pending review in https://github.com/NixOS/nixpkgs/pull/220592

**Comment by emillon** [comment]
OK, this must work already then. Thanks

**Comment by Alizter** [comment]
The slim shell works today, but the default dev shell doesn't but I didn't work out why in detail, that is still in my todo.

**Comment by rgrinberg** [comment]
I wonder if this is broken for other package managers that muck have their own installation flags. Thinking about this again, perhaps this issue is self imposed by dune itself? Suppose we were to change `dune install` to rewrite `dune-package` files on the fly to use the resolved absolute paths from the installation flags, wouldn't this issue just go away everywhere for good? The only thing that we might lose on is the reloctability of dune installations, but does it really matter?

**Comment by Alizter** [comment]
I believe this can be closed now.
